### ISSUE TITLE
General Discussion Thread

### ISSUE THREAD

**Issue by AttarGuitar** [RainbowUnicorn7297/thumper-modding-tool#3]
This is the hash function I currently have from the wiki.

uint32_t hash32(std::string s) {
    uint32_t h = 0x811c9dc5;
    for (std::string::size_type i = 0; i < s.length(); i++)
        h = (h ^ s[i]) * 0x1000193;
    h *= 0x2001;
    h = h ^ (h >> 0x7);
    h *= 0x9;
    h = h ^ (h >> 0x11);
    h *= 0x21;
    return h;
}

Unfortunately, this function only gives us the hashed parameter names, not file names. Currently, I'm trying to find .flow files as there isn't any on the wiki (We're trying to find other flow objects like crakhed_colors.flow and we can use them, however we have to use the hashed magic numbers instead of the real parameters from the game. Any advice you could give us? 
Long story short, we need to track down the original parameter paths, and the only way we think we can do it is by finding the original objlib file (for example french_horn_chords.flow from global/french_horn_chords.objlib, how would we locate french_horn_chords.objlib is the question?) and getting the parameters from there. Unfortunately stating an issue with the editor is the only way I can contact you. Any help? Much appreciated for the last fix!

**Comment by RainbowUnicorn7297** [comment]
When you use a .flow object in a .leaf or .lvl object, the associated "param_path_hash" is the magic number of the name of an external input script node. It is usually not reversible because there are no traces of the original name in either the .pc game files or the game executable. The only way to reverse it is to guess the original name and see if the string hashes into the same magic number.

**Comment by AttarGuitar** [comment]
Ah, that sucks.
We did have a guy named Biendeo on our discord actually attempt to write a bruteforce and one step was successful, but never worked in the end. 
I guess we have another question we could have explained: In all custom levels, the track has an associated rgba color on the track lights. After attempting to find out how to change the color of the track before/during the level, all attempts have been unsuccessful, even attempting to write in kTraitColor. Since none of the colors in the config file actually work for the level, we're stuck. It seems the custom levels you setup in the release all have a color set and we can't figure it out. As far as I've went, there's no evidence of a track color change in 9-27 either.

**Comment by RainbowUnicorn7297** [comment]
Unlike the level select screen, the rail colors inside a level are set in a .flow object within the level .objlib itself. They can only be changed when I add support for .flow objects to the modding tool. It's going to take a very long time, since .flow objects are the most complicated object type. There's [an example in the wiki](https://rainbowunicorn7297.github.io/#ducker_ring_sequin.objlib%3A%20public.flow) to give you an idea of what a .flow object looks like.

**Comment by AttarGuitar** [comment]
Got another question;
I'm currently developing a mid-length level, and I want to use a flow object called "dissonant_bursts.flow". However, the parameter paths listed on the wiki do not work. I think it's looking for an objlib it can't find. My question is, do I need to somehow reference it in an external file? Somehow point the game towards an objlib if that's possible?

**Comment by RainbowUnicorn7297** [comment]
"dissonant_bursts.flow" can be used directly because the custom levels references "global.objlib" in the global library list, and "global.objlib" references "global/dissonant_bursts.objlib". It should work as long as the sequencer object is formatted correctly in your custom level.

**Comment by AttarGuitar** [comment]
Got another question for you.

Let's say we lost the wiki and we had to find a file: The current hash function only hashes into parameter paths for objects, unfortunately not file names. For example, 30291331 is level_9a.objlib, right? Well, what if we had to find that file without the wiki? Due to the hash function only doing parameter paths, we couldn't find it if we wanted. Is there a different hash function?

**Comment by RainbowUnicorn7297** [comment]
For .objlib files, the cache filename is generated by applying the hash function on the lowercase of the file path with a capital 'A' prepended to it. So for level_9a.objlib, the whole thing before the hash function is "Alevels/level9/level_9a.objlib", and it becomes "30291331.pc" afterwards.
Some of the other file types use the same method, but some file types such as .x mesh files are a little bit more nuanced.

**Comment by AttarGuitar** [comment]
Got another question.
I recently learned that it is possible to define "LevelLib" in the editor. Does this mean it is possible to use leaf and lvl files from the official levels in the game? If that's possible, that would be a huge breakthrough for us.

**Comment by RainbowUnicorn7297** [comment]
While "LevelLib" is one of the object types, the game has never referenced a "LevelLib" as a global library or an external object from within. Even if you can reference .leaf or .lvl objects from the the official levels as external objects, I don't see how it's a huge breakthrough when you are only reusing the official objects.
